src/Databases/DatabaseFilesystem.cpp

```diff
@@ -0,0 +1,189 @@
+#include <Databases/DatabaseFilesystem.h>
+
+#include <IO/Operators.h>
+#include <IO/WriteBufferFromString.h>
+#include <Interpreters/Context.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ParserCreateQuery.h>
+#include <Parsers/parseQuery.h>
+#include <Storages/IStorage.h>
+#include <TableFunctions/TableFunctionFactory.h>
+
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int UNKNOWN_TABLE;
+    extern const int DATABASE_ACCESS_DENIED;
+    extern const int BAD_ARGUMENTS;
+}
+
+DatabaseFilesystem::DatabaseFilesystem(const String & name_, const String & path_, ContextPtr context_)
+    : IDatabase(name_), WithContext(context_->getGlobalContext()), path(path_), log(&Poco::Logger::get("DatabaseFileSystem(" + name_ + ")"))
+{
+    fs::path user_files_path;
+    if (context_->getApplicationType() != Context::ApplicationType::LOCAL)
+        user_files_path = fs::canonical(fs::path(getContext()->getUserFilesPath()));
+
+    if (fs::path(path).is_relative())
+        path = user_files_path / path;
+
+    path = fs::absolute(path).lexically_normal().string();
+}
+
+std::string DatabaseFilesystem::getTablePath(const std::string & table_name) const
+{
```

Comment:
let's add a check that the `path` which is in table_name is not absolute, because as I remember `fs::path(path) / table_name` will result in just `table_name` path if the path is absolute, which will mean that the user will be able to get informations about files which are outside of the scope of allowed paths. 

Would be also good to add tests with paths which go outside of allowed directory and except that it will get an error (if you add `-- { serverError <error_code>}` to a test line of sql test file, which will make sure that query returns an error as expected)

Reply:
I have already added such tests: with relative paths and absolute. See `tests/queries/0_stateless/02722_database_filesystem.sh`

Checking that the path starts with user_files already exists inside `tryGetTable`, but it is not in `isTableExists`, I will add

Reply:
done
All exceptions are checked in tests with comments